File: backend/src/main/java/io/tingkai/money/entity/ExchangeRateRecord.java
package io.tingkai.money.entity;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

import io.tingkai.money.constant.DatabaseConstant;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import lombok.Data;

@Entity
@Data
@Table(name = DatabaseConstant.TABLE_EXCHANGE_RATE_RECORD, uniqueConstraints = { @UniqueConstraint(columnNames = { "currency", "date" }) })
public class ExchangeRateRecord {
	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	protected UUID id;
	@Column(unique = true)
	protected String currency;
	protected LocalDateTime date;
	protected BigDecimal cashBuy;
	protected BigDecimal cashSell;
	protected BigDecimal spotBuy;
	protected BigDecimal spotSell;
}

File: python/src/entity/ExchangeRateRecord.py
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.schema import UniqueConstraint
import uuid

from entity.BaseEntity import BaseEntity, db

class ExchangeRateRecord(BaseEntity):
    __tablename__ = 'exchange_rate_record'

    id = db.Column(UUID(as_uuid = True), primary_key = True, default = uuid.uuid4)
    currency = db.Column(db.String, nullable = False)
    date = db.Column(db.DateTime, nullable = False)
    cash_buy = db.Column(db.Numeric, nullable = False)
    cash_sell = db.Column(db.Numeric, nullable = False)
    spot_buy = db.Column(db.Numeric, nullable = False)
    spot_sell = db.Column(db.Numeric, nullable = False)

    UniqueConstraint('currency', 'date', name = 'exchange_rate_record_currency_date_key')
